lib/filesystem.nix

```diff
@@ -76,4 +76,48 @@ in
       dir + "/${name}"
   ) (builtins.readDir dir));
 
+  # findBySuffix: String -> Path -> [ Path ]
+  #
+  # Given a suffix and directory, return a filtered list containing the full
+  # path to all files matching the given suffix recursively.
+  #
+  # NOTE: If the path parameter is a derivation output, the derivation must be built
+  #       before evaluation can proceed. This is commonly called "import from derivation"
+  #       or IFD. Unlike a normal dependency, this creates a sequential bottleneck.
+  #       If possible, process such directories in a build script instead of the Nix language.
+  #       Nixpkgs disallows import from derivation. If the path parameter is a local directory,
+  #       this note does not apply.
+  #
+  # Example:
+  #   findSuffix ".h" ./src
+  #
+  # Returns:
+  #  [ "./src/foo.h" ]
+  findSuffix = suffix: dir: (builtins.filter
+    (x: (lib.hasSuffix suffix (toString x)))
+      (lib.filesystem.listFilesRecursive dir)
+  );
+
+  # find: String -> Path -> [ Path ]
+  #
+  # Given a POSIX regular expression and directory, return a filtered list containing the full
+  # path to all files matching the given regex recursively.
+  #
+  # NOTE: If the path parameter is a derivation output, the derivation must be built
+  #       before evaluation can proceed. This is commonly called "import from derivation"
+  #       or IFD. Unlike a normal dependency, this creates a sequential bottleneck.
+  #       If possible, process such directories in a build script instead of the Nix language.
+  #       Nixpkgs disallows import from derivation. If the path parameter is a local directory,
+  #       this note does not apply.
+  #
+  # Example:
+  #   find ".+\\.h" ./src
+  #
+  # Returns:
+  #  [ "./src/foo.h" ]
+  find = pattern: dir: (builtins.filter
+    (x: ((builtins.match pattern (toString x)) != null))
+      (lib.filesystem.listFilesRecursive dir)
+  );
```

Comment:
Is there a need for this function?

Reply:
@roberth for those that like regexp, rather than suffixes.

Reply:
Regular expressions are not a good match for file system queries.

Regular expressions, as a concept, are good for matching a language: parsing a sequence of symbols into a meaningful structure. Regular expressions as implemented are generally horrible even at this because they aren't capable of returning the structure nicely. `builtins.match` inherits this flaw. Now, the structure of paths is generally quite simple, so we don't even have to invoke the concept of regular expressions in order to understand paths.

Instead, the suggestion seems to be that we can treat paths as strings, but this is error prone, because strings aren't paths. A better idea would be to generalize the concept of regular expressions to work on paths instead of strings. This is possible, because paths are still sequences. For `builtins.match`, the type of symbol of the regular expression is a character, because strings are sequences of characters. For a path matching regular expression, the type of symbol is a string, or to be specific, a string without any `/` and excluding the string `".."` (as we're discussing the paths required to represent a plain file system tree, not any path). While regular expressions are often treated as matching symbols by equality, this is usually generalized in practice, such as with character classes. In the case of paths, the common methods for matching are usually either equality (e.g. `== "flake.nix"`) or file extensions (e.g. `hasSuffix ".nix"`), but a more general means of matching might come in handy at times. This could be a regular expression. Did we come full circle? No. This is a regular expression on path components, not on paths. So our shape is not a circle, but rather two layers of regular expression, matching the structure of its inputs: a _list_ of _list_ of character.

We don't have a stringly domain specific language for such regular expressions, and this is probably a good thing. Instead, such a regular expression could be represented in a structured way, an embedded language. This would resemble a (simple) parsing library.

Also, you did't seem to need it yourself.